STOR-1065: Publish ClusterRoles for csi driver sidecars

mpatlasov opened the pull request:
The PR adds a bunch of new ClusterRoles to `manifests/` dir. This way those new ClusterRoles will be always created, regardless of cloud. They are building blocks to compose ClusterRoles for CSI dirver sidecars. For example, `external-provisioner` sidecar for `aws-ebs` csi driver can compose the same ClusterRole as in https://github.com/openshift/aws-ebs-csi-driver-operator/blob/master/assets/rbac/provisioner_role.yaml by adding ClusterRoleBindings for `openshift-csi-main-provisioner-role` and `openshift-csi-provisioner-volumesnapshot-reader-role`. The only exception is `leases` rules which need to be moved from ClusterRoles to per-namespace Roles anyway.

As soon as this change is merged into `cluster-storage-operator`, it will be possible to get rid of ClusterRole definitions in csi driver operators, they will only define ClusterRoleBindings referring these new ClusterRoles.

mpatlasov commented:
/assign @gnufied 
/label docs-approved
/label px-approved

mpatlasov commented:
/retest

mpatlasov commented on pkg/operator/csidriveroperator/csioperatorclient/aws.go:
```diff
@@ -46,11 +48,15 @@ func GetAWSEBSCSIOperatorConfig(isHypershift bool) CSIOperatorConfig {
 			"csidriveroperators/aws-ebs/hypershift/mgmt/02_sa.yaml",
 			"csidriveroperators/aws-ebs/hypershift/mgmt/04_rolebinding.yaml",
+			"sidecars/main_provisioner_role.yaml",
+			"sidecars/volumesnapshot_provisioner_role.yaml",
```
We clarified this topic in a chat with Hemant. For posterity: an instance of csi driver running inside Management Cluster communicates with Guest Cluster API server, so ClusterRoles published there (in `csiDriverConfig.GuestStaticAssets` above) would fully suit its needs.

jsafrane commented on assets/sidecars/lease_provisioner_role.yaml:
```diff
@@ -0,0 +1,8 @@
+kind: ClusterRole
+apiVersion: rbac.authorization.k8s.io/v1
+metadata:
+  name: lease-external-provisioner-role
+rules:
+  - apiGroups: ["coordination.k8s.io"]
+    resources: ["leases"]
+    verbs: ["get", "watch", "list", "delete", "update", "create"]
```
IMO this is not related to provisioning at all. I'd use some global-sounding role (`csi-driver-lease-leader-election`?) and use just RoleBinding (not ClusterRoleBinding), because the sidecars need these perms. only in namespaces where they run.

mpatlasov replied:
> IMO this is not related to provisioning at all.

We currently have 3 operators defining this role:
https://github.com/openshift/aws-efs-csi-driver-operator/blob/master/assets/rbac/provisioner_role.yaml#L25-L26
https://github.com/openshift/azure-file-csi-driver-operator/blob/main/assets/rbac/provisioner_role.yaml#L25-L26
https://github.com/openshift/ibm-powervs-block-csi-driver-operator/blob/main/assets/rbac/provisioner_role.yaml#L31-L32

jsafrane commented on assets/sidecars/volumesnapshot_provisioner_role.yaml:
```diff
@@ -0,0 +1,11 @@
+kind: ClusterRole
+apiVersion: rbac.authorization.k8s.io/v1
+metadata:
+  name: volumesnapshot-external-provisioner-role
+rules:
+  - apiGroups: ["snapshot.storage.k8s.io"]
+    resources: ["volumesnapshots"]
+    verbs: ["get", "list"]
+  - apiGroups: ["snapshot.storage.k8s.io"]
+    resources: ["volumesnapshotcontents"]
+    verbs: ["get", "list"]
```
dtto, if a storage backed supports snapshots, this (and more) will be part of external-snapshotter role.

mpatlasov replied:
Same as in the comment above: provisioner needs only two verbs: `["get", "list"]`, while snapshotter wants more: `["get", "list", "watch", "update"]`.

mpatlasov commented on assets/sidecars/configmap_provisioner_role.yaml:
```diff
@@ -0,0 +1,8 @@
+kind: ClusterRole
+apiVersion: rbac.authorization.k8s.io/v1
+metadata:
+  name: configmap-external-provisioner-role
+rules:
+  - apiGroups: [""]
+    resources: ["configmaps"]
+    verbs: ["get", "list"]
```
https://github.com/openshift/ibm-vpc-block-csi-driver-operator/blob/master/assets/rbac/provisioner_role.yaml#L87-L90

mpatlasov replied:
@dobsonj You were involved in https://github.com/openshift/ibm-vpc-block-csi-driver-operator/pull/40, and [this comment](https://github.com/openshift/ibm-vpc-block-csi-driver-operator/pull/40/commits/ff4825c94286a4bb9cdcc3d74442fb0e0b3fd6f1#r901363637):

> adding the resources and permissions for provisioner role should suffice. It can be removed from resizer and registrar.

suggests that "configmaps" rule was added intentionally. Do you remember why `external-provisioner` needs it and what makes `ibm-vpc-block-csi-driver-operator` so special that other operators are happy w/o it?

dobsonj replied:
AFAIK, [this configmap](https://github.com/openshift/ibm-vpc-block-csi-driver-operator/blob/master/assets/configmap.yaml) is needed by the CSI driver container, but I can't think of a reason why external-provisioner would need it.

mpatlasov commented on pkg/operator/csidriveroperator/csioperatorclient/alibaba-disk.go:
```diff
@@ -29,6 +29,8 @@ func GetAlibabaDiskCSIOperatorConfig() CSIOperatorConfig {
 			"csidriveroperators/alibaba-disk/04_rolebinding.yaml",
 			"csidriveroperators/alibaba-disk/05_clusterrole.yaml",
 			"csidriveroperators/alibaba-disk/06_clusterrolebinding.yaml",
+			"sidecars/main_provisioner_role.yaml",
+			"sidecars/volumesnapshot_provisioner_role.yaml",
```
I like this idea. I'll try to move all roles (for all sidecars AND all operators) into `manifests/` dir under some global-sounding names like `openshift-csi-lease-leader-election-role` and `openshift-csi-secret-reader-role`.

mpatlasov commented on assets/sidecars/volumeattachment_provisioner_role.yaml:
```diff
@@ -0,0 +1,8 @@
+kind: ClusterRole
+apiVersion: rbac.authorization.k8s.io/v1
+metadata:
+  name: volumeattachment-external-provisioner-role
+rules:
+  - apiGroups: ["storage.k8s.io"]
+    resources: ["volumeattachments"]
+    verbs: ["get", "list", "watch"]
```
We have to distinguish them because "verbs" lists are different: `["get", "list", "watch"]` for provisioner, and `["get", "list", "watch", "update", "patch"]` for attacher. It would impact security if we let provisioner to modify volumeattachments although it needs only read-only access to them.

jsafrane replied:
The provisioner and attacher use the same service account, so there is no impact on security. IMO having the permissions that detailed will impact readability.

mpatlasov replied:
@jsafrane , how can we apply your explanation about provisioner and attacher to the manila usecase? We don't have `attacher_role.yaml` there, but in `provisioner_role.yaml`:

```
  - apiGroups: ["storage.k8s.io"]
    resources: ["volumeattachments"]
    verbs: ["get", "list", "watch"]
```
(https://github.com/openshift/csi-driver-manila-operator/blob/master/assets/rbac/provisioner_role.yaml#L24-L26)

jsafrane replied:
It looks like a bug in Manila then.

jsafrane replied:
or, maybe not...  Keep the reader role there, we'll see when we get to individual drivers.

mpatlasov commented on assets/sidecars/secret_provisioner_role.yaml:
```diff
@@ -0,0 +1,8 @@
+kind: ClusterRole
+apiVersion: rbac.authorization.k8s.io/v1
+metadata:
+  name: secret-external-provisioner-role
+rules:
+  - apiGroups: [""]
+    resources: ["secrets"]
+    verbs: ["get", "list", "watch"]
```
verbs lists are slightly different: `["get", "list", "watch"]` for provisioner, and only `["get", "list"]` for snapshotter. I think it's OK to simplify things having a single `openshift-csi-secret-reader.role` with `["get", "list", "watch"]` for all sidecars who wants it.

@dobsonj , from "Threat Model" standpoint, are we supposed to distinguish between `["get", "list"]` and `["get", "list", "watch"]`?

dobsonj replied:
IMO, I think it's acceptable to create a single read-only role (including get/list/watch) for all sidecars to use.

mpatlasov commented:
The last [force-push](https://github.com/openshift/cluster-storage-operator/pull/379#event-9453548956) fully reworks PR putting all ClusterRole building blocks (for all sidecars) to `manifests/` dir. See this [comment](https://github.com/openshift/cluster-storage-operator/pull/379#discussion_r1219542371).

mpatlasov commented:
/retest

mpatlasov commented:
/retest

mpatlasov commented:
New force-push updates PR:
1) remove `manifests/openshift-csi-csinodeinfo-reader-role.yaml` altogether (per discussion with Jan)
2) added annotations to make CVO happy

mpatlasov commented:
/retest ci/prow/hypershift-aws-e2e-external

mpatlasov commented:
/test ci/prow/hypershift-aws-e2e-external

mpatlasov commented:
/test hypershift-aws-e2e-external

mpatlasov commented:
/retest

jsafrane commented on manifests/sidecar-main_attacher_role.yaml:
```diff
@@ -0,0 +1,16 @@
+    verbs: ["get", "list", "watch", "update", "patch"]
+  - apiGroups: [""]
+    resources: ["nodes"]
+    verbs: ["get", "list", "watch"]
```
I would expect some rules for `volumeattachment` - the attacher reads/writes them a lot

mpatlasov replied:
Yes, that's right. The attacher will bind to openshift-csi-volumeattachment-reader-role and openshift-csi-volumeattachment-writer-role. I didn't include these rules into `openshift-csi-main-attacher-role` to avoid verb duplication.

mpatlasov commented:
/retest

mpatlasov commented:
New force-push updates PR:
 * add a rule for `csinodes` to `sidecar-main_attacher_role.yaml`
 * remove `openshift-csi-csinode-reader-role.yaml` altogether

This simplifies the code because all `attacher` sidecars need prems to read csinodes.

mpatlasov commented:
/test hypershift-aws-e2e-external

mpatlasov commented:
New force-push updates PR:
 * added rules to `manifests/sidecar-main_attacher_role.yaml`. Now it suites all operators
 * reworked `manifests/openshift-csi-*` files to correspond to specific sidecars
 
 (see https://docs.google.com/document/d/1oiGN7ulFM3BE-fPb6U2mbSWbAQCJYyFJm3CYzHrkKkY/edit?usp=sharing)

mpatlasov commented:
/test hypershift-aws-e2e-external

mpatlasov commented:
/test e2e-azure-csi

mpatlasov commented:
This PR passes e2e tests in conjunction with [STOR-1065: Publish ClusterRoles for csi driver sidecars](https://github.com/openshift/cluster-storage-operator/pull/379):
```
job <https://prow.ci.openshift.org/view/gs/origin-ci-test/logs/release-openshift-origin-installer-launch-aws-modern/1671702027480076288 | test e2e https://github.com/openshift/cluster-storage-operator/pull/379,https://github.com/openshift/aws-ebs-csi-driver-operator/pull/244 aws> succeeded
```
(https://prow.ci.openshift.org/view/gs/origin-ci-test/logs/release-openshift-origin-installer-launch-aws-modern/1671702027480076288)

jsafrane commented on manifests/openshift-csi-resizer-storageclass-reader-role.yaml:
```diff
@@ -0,0 +1,13 @@
+kind: ClusterRole
+apiVersion: rbac.authorization.k8s.io/v1
+metadata:
+  name: openshift-csi-resizer-storageclass-reader-role
```
This role is suspicious - is there a resizer that does not need to read StorageClasses?

jsafrane commented on manifests/sidecar-main_registrar_role.yaml:
```diff
@@ -0,0 +1,32 @@
+kind: ClusterRole
+apiVersion: rbac.authorization.k8s.io/v1
+metadata:
+  name: openshift-csi-main-driver-registrar-role
```
In https://github.com/openshift/ibm-vpc-block-csi-driver-operator/pull/63 I checked that these rules are ibm-vpc-node-label-updater rules, not registrar's. Node-driver-registrar does not talk to the API server at all.

This ClusterRole (or Role?) belongs to IBM VPC CSI driver operator only.

mpatlasov replied:
OK, `manifests/sidecar-main_registrar_role.yaml` is removed!

jsafrane reviewed (commented):
I am fine if it's merged as it is, we can always fix it later. Still, there is one issue below.

mpatlasov commented:
/test e2e-ovn-vsphere
/test e2e-vsphere-csi

mpatlasov commented:
/test e2e-azure-csi
/test e2e-azure-ovn

gnufied commented:
/lgtm
/approve
